spring-kafka/src/test/java/org/springframework/kafka/core/DefaultShareConsumerFactoryTests.java

```diff
@@ -197,4 +185,102 @@ void integrationTestDefaultShareConsumerFactory(EmbeddedKafkaBroker broker) thro
+				recordCount, broker);
+
+		// Assert all records were received (no loss)
+		Set<String> allReceived = new java.util.HashSet<>();
+		consumerRecords.values().forEach(allReceived::addAll);
+		for (int i = 0; i < recordCount; i++) {
+			assertThat(allReceived)
+				.as("Should have received value " + topic + "-value-" + i)
+				.contains(topic + "-value-" + i);
+		}
+	}
+
+	/**
+	 * Runs multiple Kafka consumers in parallel using ExecutorService, collects all records received,
+	 * and returns a map of consumerId to the set of record values received by that consumer.
+	 */
+	private static Map<String, Set<String>> runSharedConsumerTest(String topic, String groupId,
+			List<String> consumerIds, int recordCount, EmbeddedKafkaBroker broker) throws Exception {
+		var bootstrapServers = broker.getBrokersAsString();
+
+		var producerProps = new java.util.Properties();
+		producerProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
+		producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+		producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+		try (var producer = new KafkaProducer<String, String>(producerProps)) {
+			for (int i = 0; i < recordCount; i++) {
+				producer.send(new ProducerRecord<>(topic, "key" + i, topic + "-value-" + i)).get();
+			}
+			producer.flush();
+		}
+
+		setShareAutoOffsetResetEarliest(bootstrapServers, groupId);
+
+		Map<String, Set<String>> consumerRecords = new java.util.concurrent.ConcurrentHashMap<>();
+		consumerIds.forEach(id -> consumerRecords.put(id,
+				java.util.Collections.synchronizedSet(new java.util.HashSet<>())));
+		var latch = new java.util.concurrent.CountDownLatch(recordCount);
+		var running = new java.util.concurrent.atomic.AtomicBoolean(true);
+		ExecutorService executor = Executors.newCachedThreadPool();
+		DefaultShareConsumerFactory<String, String> shareConsumerFactory = new DefaultShareConsumerFactory<>(
+				Map.of(
+					"bootstrap.servers", bootstrapServers,
+					"key.deserializer", org.apache.kafka.common.serialization.StringDeserializer.class,
+					"value.deserializer", org.apache.kafka.common.serialization.StringDeserializer.class
+				)
+		);
+		for (int i = 0; i < consumerIds.size(); i++) {
+			final int idx = i;
+			executor.submit(() -> {
+				try (var consumer = shareConsumerFactory.createShareConsumer(groupId, consumerIds.get(idx))) {
+					consumer.subscribe(Collections.singletonList(topic));
+					while (running.get() && latch.getCount() > 0) {
+						var records = consumer.poll(Duration.ofMillis(200));
+						for (var r : records) {
+							if (consumerRecords.get(consumerIds.get(idx)).add(r.value())) {
+								consumer.acknowledge(r, AcknowledgeType.ACCEPT);
+								latch.countDown();
+							}
+						}
+					}
+				}
+			});
+		}
+
+		assertThat(latch.await(10, TimeUnit.SECONDS))
+			.as("All records should be received within timeout")
+			.isTrue();
+		running.set(false);
```

Comment:
Can we instead of this flag really close those consumers after that latch is fulfilled?